### ISSUE TITLE
Support AoE2DE

### ISSUE THREAD

**Issue by krikru** [kachayev/pyage2#1]
Hi!

I play a decent amount of _Age of Empires II: Definitive Edition_, and it would be awesome to be able to train bots to play it, so I think this project seems really interesting. Would it be possible to make it support AoE2DE as well?

**Comment by kachayev** [comment]
Hi @krikru! Thanks a lot for expressing your interest. Unfortunately, there's no public API to make proper integration with the game. Thus everything works based on a few tricks and hack which is sensitive to  any changes in the game code, e.g. between patches or versions.

There's one thing that could theoretically work but I've never tried the implementation myself: theoretically it should be possible to generate the script for the bot (it's in lisp-like) and run the normal game against "normal" bots. This is a very different setup from RL point of view but it might be an appealing research avenue.

**Comment by krikru** [comment]
Hi @kachayev, okay, I was hoping that it would be a more stable solution, but I guess that's always wishful thinking when there is no official support for that kind of things. I think then this wasn't really what I was looking for, even though I still think the idea is very interesting. I'll cross my fingers that the very unlikely thing happens that Microsoft add such support, and please notify me if you at some point notice that they do!

It's true that you could generate a bot script and run it like a normal bot, and it would be very interesting to see that done, but I have no idea how you would do that. In my view, you would either need to 1) use reinforcement learning directly on the bot script, which it is not obvious to me how you would do (even though similar things have been done with behavior trees, which I guess [these](https://openaccess.city.ac.uk/id/eprint/3000/) [papers](https://arxiv.org/abs/1504.05811) [describe](https://arxiv.org/abs/1506.02312), but I'm really no expert on the subject), or 2) you train a RL agent like "normal" using deep learning (this would still require unofficial tools like yours to work) and then you convert the resulting policy network to a script (which I don't know how to do, but would be cool to know if there would be some way to do it). I think that out of these two methods, the second alternative still seems like like the most appealing, as I have a feeling that you may get more human-like behaviors with deep RL that if you learn the script directly without any neural network. But I really don't know for sure and both would be very interesting to see tried out.

**Comment by kachayev** [comment]
I don't know if there's a connection between the underlying architecture and "human-likeness", it's an interesting question. Intuitively, physiology suggests that our decision-making process is a plan-ahead with a few variables/conditions. Which would be closer to the script with regularized width and depth of each branch. As far as I know, there's no existing solution to do planning/search with NN. I remember MSR did research on learning human-like navigation with RL, that might give some insights into the problem, though navigation is much lower-level skill compared to strategic decision making.

Regarding learning, I would imagine that evolution strategy (ES) with a mutation operators designed for the decision tree might work out rather nicely. Maybe even taking existing algorithms would do, just to plug defined mutation and see how it works. Better to try on a smaller game, for sure. This one is computationally intense.

**Comment by krikru** [comment]
I'm thinking that with NNs, you naturally get decision boundaries that are rather flexible, given the large degree of freedom of the NNs, and that don't have to be axis-aligned (in input space), which I imagine is closer to how human decision boundaries look like. But granted, I don't know how AoE2's AI script work (they may allow for very flexible decision boundaries) and I don't really know what human decision boundaries look like either.

Regarding planning/searching with NN, isn't that what board game engines like AlphaZero and MuZero do? Some have even speculated that this is what the mysterious Q* algorithm by OpenAI does too.

I'm personally hesitant towards evolutionary algorithms. I feel like there has to be better learning algorithms. When a human improves their strategy on some task, unlike evolutionary algorithms, the don't just change arbitrary things in their strategy, but they change things that they know (or suspect might) have a strong influence on the thing they want to improve, which is kind of what NNs trained with backpropagation do too and which I think is probably a more efficient way to try new strategies. I just know that deep RL trained with backpropagation has led to some phenomenal results in many different types of board games and video games (including StarCraft II which should be quite similar to AoE2) and I haven't really seen any other type of ML come close in performance. But some research groups have suggested evolutionary strategies to solve RL problems before (including [Schmidhuber](https://people.idsia.ch/~juergen/evolution.html) and [OpenAI](https://arxiv.org/abs/1703.03864)), and I would definitely have a look if I saw something that seemed to be able to rival gradient-based learning.

**Comment by kachayev** [comment]
> Regarding planning/searching with NN, isn't that what board game engines like AlphaZero and MuZero do?

RL agents primarily employ NNs for three purposes: predicting the value of a given state (so-called "value network"), forecasting the transition from a current state and action (as 'model' in model-based RL), and determining the next action from a given state (so-called "policy" in model-free RL). Notably, these functions do not constitute 'planning' in the traditional sense. When planning is incorporated into an RL agent (e.g. any model-based or simulation-based or "mixed" algorithms), it typically treats the NN (or a simulator) as a black-box. For example, AlphaZero uses the Monte Carlo Tree Search (MCTS) algorithm for this purpose. The concept of using NNs specifically for planning was once a popular topic. However, this interest faded away as time passed by, with the prevailing view shifting towards the need for a recurrent closed-loop control system in such applications.

> to rival gradient-based learning

By default, RL environments are assumed to be non-differentiable, positioning RL as a black-box optimization method particularly suited for such environments. This assumption aligns RL with the types of problems tackled by ES. Environments built with components like end-to-end differentiable physics engines are rare exceptions. Differentiable objectives in this field are developed as proxies, which solve certain problems but introduce new ones. Curiously, the most advanced learning mechanism known to us – the human brain – is unlikely to employ gradients in its learning processes. Thus I would hesitate to count non-gradient methods out of the race. Even though the most money and energy are dedicated to the former nowadays.

**Comment by krikru** [comment]
> The concept of using NNs specifically for planning was once a popular topic.

Hm, I meant that AlphaZero and MuZero did planning/search with NNs in the sense that they did planning and used NNs in the process, even though admittedly not for everything. How do you mean that you could do planning with NNs instead? I'm intrigued. I understand that the way it's performed in MuZero is not ideal in every situation, probably for many reasons, but one specific reason is that I guess it assumes a fixed timestep in the search tree, and you might want to be able to adapt the step length or just jump straight to when you know that something interesting is going to happen which you can't do with MuZero.

> By default, RL environments are assumed to be non-differentiable, positioning RL as a black-box optimization method particularly suited for such environments. This assumption aligns RL with the types of problems tackled by ES. Environments built with components like end-to-end differentiable physics engines are rare exceptions. Differentiable objectives in this field are developed as proxies, which solve certain problems but introduce new ones.

I don't think the environment itself needs to be differentiable for gradient-based methods to make sense, since the neural networks will learn a model of the environment, which in turn can be differentiated. But even if that model wouldn't be differentiable with respect to the game state either, it would still be differentiable with respect to its parameters, which is enough to train it.

Also, when you use gradient-based learning, there is a bunch of ways to implement artificial curiosity (including probably a couple that I would like to try out if I got the opportunity, but haven't seen anyone else try out), which I personally find very interesting, and I think this is another important key for solving hard reinforcement learning problems. but I'm not really sure if it is possible to implement curiosity in an agent that you train with an evolutionary algorithm (EA; but the need for doing so is probably not as big either, since they don't have a problem with sparse reward signals).

> Thus I would hesitate to count non-gradient methods out of the race. Even though the most money and energy are dedicated to the former nowadays.

You're certainly right, and I don't know how good EAs are; maybe they are just as good as gradient-based algorithms in many cases and maybe even better in some cases. It would be interesting to see them used more often. I'm also probably quite biased because of the current trend in ML research.

**Comment by kachayev** [comment]
> How do you mean that you could do planning with NNs instead?

There are at least a few angles this question could be tackled. E.g. you can think about forward pass of the NN performing computation thus it might be theoretically possible to find weights so that this computation coincides with the search. Or, you can formalize search as a function that outputs trajectory and to use NN as the approximation. Or, you can use RNN state the same way you would keep state between steps in breadth-first tree scan. The biggest problem in all of those is how to isotopically embed exponentially growing state space on the each step into a finite dimensional Euclidean. Plus typical solution set is a polytope which, again, won't have gradients exactly where you need them the most.

**Comment by kachayev** [comment]
> I don't think the environment itself needs to be differentiable for gradient-based methods to make sense, since the neural networks will learn a model

Yes, absolutely. That's what I meant by "proxies". But from math point, if your optimization objective is convex and smooth, you can proof that gradient leads to optimal convergence rate. If not... then all bets are off. RL objective - expectation of future return - is not differentiatable w.r.t. to policy if environment is not (i.e it's a black-box).

I agree that there are some parts of the story where gradient-based learning for the NN-parametrized solution fits perfectly. Those are: 1) feature extractor from high-dimensional state space (e.g. from pixels) - could be learned as an auto encoder with any suitable loss; 2) so-called "world model" could be trained to predict transition merely as an auto regressive model, again with any suitable loss. Note that both are not "RL" task per se, they give you building blocks that you can use to solve RL problem but overall are tangential to the original objective from optimization perspective.

When it comes to the "RL" part, there were at least a few works demonstrating that Deep RL remains brittle even on simple tasks, and may fail where simplest optimal control models show high performance.

What am I trying to say, is that RL is a complex field, and we don't have (and we might never have) a definitive "X is all you need" type of situation. Especially if we want RL to be used broadly in practice.
